src/lib/structures/Argument.ts

```diff
@@ -116,10 +117,106 @@ export abstract class Argument<T = unknown> extends BaseAliasPiece implements IA
 	}
 }
 
+/**
+ * The extended argument class. This class is abstract and is to be extended by subclasses which
+ * will implement the [[ExtendedArgument#handle]] method.
+ * Much like the [[Argument]] class, this class handles parsing user-specified command arguments
+ * into typed command parameters. However, this class can be used to expand upon an existing
+ * argument in order to process its transformed value rather than just the argument string.
+ *
+ * @example
+ * ```typescript
+ * // TypeScript:
+ * import { ApplyOptions } from '@sapphire/decorators';
+ * import { ArgumentResult, ExtendedArgument, ExtendedArgumentContext, ExtendedArgumentOptions } from '@sapphire/framework';
+ * import type { Channel, TextChannel } from 'discord.js';
+ *
+ * // Just like with `Argument`, you can use `export default` or `export =` too.
+ * @ApplyOptions<ExtendedArgumentOptions>({
+ *   name: 'textChannel',
+ *   baseArgument: 'channel'
+ * })
+ * export class TextChannelArgument extends ExtendedArgument<'channel', TextChannel> {
+ *   public handle(parsed: Channel, { argument }: ExtendedArgumentContext): ArgumentResult<TextChannel> {
+ *     return parsed.type === 'text'
+ *       ? this.ok(parsed as TextChannel)
+ *       : this.error(argument, 'ArgumentTextChannelInvalidTextChannel', 'The argument did not resolve to a text channel.');
+ *   }
+ * }
+ * ```
+ *
+ * @example
+ * ```javascript
+ * // JavaScript:
+ * const { ExtendedArgument } = require('@sapphire/framework');
+ *
+ * module.exports = class TextChannelArgument extends ExtendedArgument {
+ *   constructor(context) {
+ *     super(context, { name: 'textChannel', baseArgument: 'channel' });
+ *   }
+ *
+ *   handle(parsed, { argument }) {
+ *     return parsed.type === 'text'
+ *       ? this.ok(parsed)
+ *       : this.error(argument, 'ArgumentTextChannelInvalidTextChannel', 'The argument did not resolve to a text channel/');
+ *   }
+ * }
+ * ```
+ */
+export abstract class ExtendedArgument<K extends keyof ArgType, T> extends Argument<T> {
```

Comment:
We try to follow one-class-per-module :)

Reply:
Should be resolved as of 8bc4a3d.